### PR TITLE
Mutable PageAcc

### PR DESCRIPTION
**Pull request by phadej**

### CODE REVIEW

**Review by dcoutts — COMMENTED** [review]
I think this is a very plausible approach, and likely to be pretty fast and have low allocation/GC overheads.

I like the neat trick that all the accumulator arrays are pre-allocated to their maximum size. So a single `MPageAcc` accumulator could be reset and reused (without new allocations) for all the output pages.

```
unsafeFreezePageAcc :: MPageAcc s -> ST s PageAcc
```
This is almost but not quite what I expected. I was assuming that this action will copy all the accumulator arrays (some with appropriate offset adjustments) into the right places in a final `RawPage`. So it wouldn't be unsafe since it would be making copies, and the result would be a `RawPage`.
```
freezePageAcc :: MPageAcc s -> ST s RawPage
```
or perhaps get given a (reusable) mutable raw page to write into
```
freezePageAcc :: MPageAcc s -> MRawPage s -> ST s ()
```

The `MPageAcc` representation contains equivalents of some of the `PageSize` counters in the form of the `mpDir` two element array.
```
data PageSize = PageSize {
                  pageSizeElems :: !Int,
                  pageSizeBlobs :: !Int,
                  pageSizeBytes :: !Int
                }
```
But the comments says it's only a 2-element array
```
-- ^ array of two Word16 for keys and blobrefs count.
```
if that were just extended to be a 3-element array with the page size in bytes, then you'd have a (mutable) representation of the `PageSize`, needed to calculate if `pageAccAddElem` can squeeze another element into the page accumulator.

The other nice thing we could do here is provide a "reset" operation that resets the counters to the right values for an empty page (which is not zero for the page size), ready to reuse the `MPageAcc` for a subsequent page. It'd also have to clear the bitmaps (or the prefix that was used) so that we don't end up with trailing junk bits in later pages.

As discussed elsewhere we don't need builder support. Producing a `RawPage` (which we can guarante to be pinned) or writing into a pre-existing allocated `MRawPage` would be enough to cover our I/O needs, by non-copying conversion from raw page to `ByteString`.

File: src/Database/LSMTree/Internal/PageAcc.hs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    P.setPrimArray mpBlobRefs1 0 680 0
+    P.setPrimArray mpBlobRefs1 0 680 0
+
+-- | Write the first entry to the 'MPageAcc'.
+_addFirstEntry
+    :: MPageAcc s
+    -> SerialisedKey
+    -> Entry SerialisedValue BlobSpan
+    -> ST s ()
+_addFirstEntry MPageAcc {..} k e = assert (sizeofEntry k e <= maxsizeofEntry) $ do
```
Comment (dcoutts) [inline src/Database/LSMTree/Internal/PageAcc.hs:135]: Hmm, do we want to separate the first from remaining entries? I know the first one can't fail to fit, but is it worth separating it out? Not clear.

**Review by dcoutts — COMMENTED** [review]
Yes this looks like it's going in the right direction. The PageAcc exported API looks right.

File: src/Database/LSMTree/Internal/PageAcc.hs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+-- | Reuse 'MPageAcc' for constructing new page, the old data will be reset.
+resetPageAcc
+    :: MPageAcc s
+    -> ST s ()
+resetPageAcc MPageAcc {..} = do
+    P.setPrimArray mpDir 0 3 0
+    P.setPrimArray mpOpMap 0 11 0
+    P.setPrimArray mpBlobRefsMap 0 22 0
+    P.setPrimArray mpBlobRefs1 0 680 0
+    P.setPrimArray mpBlobRefs1 0 680 0
```
[the remark this answers is not in the document]
Comment (phadej) [reply src/Database/LSMTree/Internal/PageAcc.hs:127]: >  we don't need to zero out these. 

We need. https://hackage.haskell.org/package/primitive-0.9.0.0/docs/Data-Primitive-PrimArray.html#v:newPrimArray

> Create a new mutable primitive array of the given length. The underlying memory is left uninitialized.

**Review by mheinzel — APPROVED** [review]
Nice. Just a few small remarks, feel free to disregard.

File: src/Database/LSMTree/Internal/PageAcc.hs [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+byteSizeIdx = 2
+{-# INLINE byteSizeIdx #-}
+
+keysSizeIdx :: Int
+keysSizeIdx = 3
+{-# INLINE keysSizeIdx #-}
+
+pageSize :: Int
+pageSize = 4096
+{-# INLINE pageSize #-}
```
[the remark this answers is not in the document]
Comment (phadej) [reply src/Database/LSMTree/Internal/PageAcc.hs:99]: Unfortunately there are no way to test that.

File: src/Database/LSMTree/Internal/PageAcc.hs [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+-- as there are only 256 single byte keys.
+--
+-- >>> let calc2 s ps = case pageSizeAddElem (Key $ if s < 257 then "x" else "xx", Delete, Nothing) ps of { Nothing -> s; Just ps' -> calc2 (s + 1) ps' }
+-- >>> calc2 1 page0
+-- 680
+--
+-- Rounded to the next multiple of 64 is 704
+--
+-- 'MPageAcc' can hold up to 704 elements, but most likely 'pageAccAddElem' will make it overflow sooner.
+-- Having an upper bound allows us to allocate all memory for the accumulator in advance.
```
[the remark this answers is not in the document]
Comment (phadej) [reply src/Database/LSMTree/Internal/PageAcc.hs:63]: good idea, as other constants are defined, these should be too. Will do.

[the remark this answers is not in the document]
Comment (phadej) [reply src/Database/LSMTree/Internal/PageAcc.hs:63]: Except no. I'd need `div64 maxItemCount`, and i'm not sure GHC will constant fold bit operations.

**Pull request merged**